### PR TITLE
Compile against recent google WebRTC snapshot for ICE support

### PR DESCRIPTION
**Pull request by ZCube**
#117 #118 #121 

## Compilation WebRTC (refs/remotes/branch-heads/4180)

Add SteamWebRTC CMake based build scripts.

- Add CMake based WebRTC build scripts (Minimalized for ICE transport).
- Add Abseil-cpp dependency for WebRTC.
- Update travis script to testing steamwebrtc build step.
- Add option to build steamwebrtc as static library.
- Update IceSession for Turn server configuration.
- Add ws2_32, crypt32 dependencies for static OpenSSL library link.
- Update WebRTC's rtc_base/system/arch.h for ppc64le, s390x build.
- Add HAVE_STRCPY_S definition for check strcpy_s

Usage:
```
cmake -DUSE_STEAMWEBRTC=ON .. 
cmake -DUSE_STEAMWEBRTC=ON -DSTEAMWEBRTC_USE_STATIC_LIBS=OFF ..
cmake -DUSE_STEAMWEBRTC=ON -DSTEAMWEBRTC_USE_STATIC_LIBS=ON ..
```

### CODE REVIEW

**Comment by fletcherdvalve** [comment]
Hey just wanted to say thanks again for doing this.  When you're done, I'll probably ask for help from @tycho to review all the build setup stuff.  I can work on setting up some sort of example / test framework / fake signaling sevrice.  I have quite a few fairly large P2P changes incoming, but I don't think they will conflict with any of this work.

**Comment by tycho** [comment]
Please make sure to clean up the commit history for this patch series. Follow the guidelines presented [here](https://github.com/torvalds/subsurface-for-dirk/blob/a48494d2fbed58c751e9b7e8fbff88582f9b2d02/README#L88) and [here](https://www.kernel.org/doc/html/v4.17/process/submitting-patches.html#the-canonical-patch-format), in order to write good commit messages. You definitely should squash several of these commits with a rebase as well. The general ordering for commit history in a good pull request is:

1. Any commits to refactor or prepare the existing code base for your changes.
2. Commits implementing your change, including appropriate build system changes (both CMake and Meson, in this project's case).
3. Any additional followup changes/bugfixes (generally these can be squashed into the commits implementing the change after they've received appropriate review).

The Linux kernel project has some [good guidelines](https://www.kernel.org/doc/html/v4.17/process/submitting-patches.html#separate-your-changes) on this.

Note as well: for each commit in sequence, the build should *not* break -- as in, a Travis build should be able to succeed at every point in the history.  This is necessary in case a revert of a specific logical change needs to happen in the future.

This contribution policy really helps with long-term code maintenance. For instance, in case something breaks it is helpful to have a clear and concise history in the commit log to help understand what was done, what was intended, and what we can do to fix it.

File: .travis/install-archlinux.sh [hunk 1]
```diff
@@ -21,4 +21,6 @@ PACKAGES+=(libsodium)
 pacman --noconfirm -Syu
 pacman --noconfirm -Sy "${PACKAGES[@]}"
 
+python3 -m ensurepip --default-pip
```
Comment (tycho) [inline .travis/install-archlinux.sh:21]: First off, I haven't seen why you've added this yet, but you should not need pip, unless there really is some Python module you depend on that doesn't exist as an OS package.

If you just need to be able to run a setup.py script, then you should be installing setuptools instead. Additionally, if you're building/testing a Python module, then you should probably be using virtualenv (which includes pip and setuptools).

If you *really* need pip, then this should just be added to the PACKAGES array for Arch as "python-pip".

Comment (ZCube) [reply .travis/install-archlinux.sh:21]: Sorry. I am still unfamiliar with using github and travis. I was working on my fork. But I forgot to create another branch.

There are additional dependencies for WebRTC. I was testing using conan, so I tried to install python and conan. There were errors in the installation, so I've tested in a number of ways.

I will also modify arch linux as a package.

The dependencies I wanted to install conan are abseil and protobuf.

An error occurred in protobuf build in arch linux, abseil was not in the distributions.

Sorry for bad writing. I'm not good at english.

**Comment by tycho** [comment]
Adding another comment so I don't lose this link... someone on Hacker News linked to a fantastic guide on writing strong git commit messages: https://chris.beams.io/posts/git-commit/

Also, your recent commit messages are looking a bit better, though the changes still need quite a bit of cleanup. (I realize it's a work in progress, though.)

**Comment by ZCube** [comment]
@tycho I clean up commits. But there are some problem on master branch now. so I couldn't pass travis builds.

**Comment by tycho** [comment]
Yes, ignore the Arch Linux builder results for now. There's a bug filed with LLVM to try and get to the bottom of that one:

https://bugs.llvm.org/show_bug.cgi?id=46313

**Comment by fletcherdvalve** [comment]
> Yes, ignore the Arch Linux builder results for now.

Is it a single-line change in some travis config to nuke that config for now?  I hate to see the build "red" for so long, it doesn't look good.....   If you can point me in the right direction I'll make the change.

**Comment by fletcherdvalve** [comment]
I will hopefully get to looking at this change this week .  Good stuff!

**Comment by ZCube** [comment]
> > Yes, ignore the Arch Linux builder results for now.
> 
> Is it a single-line change in some travis config to nuke that config for now? I hate to see the build "red" for so long, it doesn't look good..... If you can point me in the right direction I'll make the change.

on fedora:rawhide
```
>echo _ZN6google8protobuf13RepeatedFieldIiEC2ERKS2_ | c++filt
google::protobuf::RepeatedField<int>::RepeatedField(google::protobuf::RepeatedField<int> const&)
>nm -gD /usr/lib/libprotobuf.so  |  grep 0000000000117b80 |  c++filt
google::protobuf::RepeatedField<int>::RepeatedField(google::protobuf::RepeatedField<int> const&)
>0000000000117b80 W _ZN6google8protobuf13RepeatedFieldIiEC1ERKS2_
```
Same function, but with a different name. I don't know the difference between these names. I think it would be better to build and use Protobuf.

**Comment by tycho** [comment]
I don't think building protobuf ourselves will help. This seems to be an LLVM compiler bug.

**Comment by fletcherdvalve** [comment]
Just a note that I haven't forgotten about this, and hopefully will act on it soon. :)

**Comment by fletcherdvalve** [comment]
Just pulled this in.  I am on bionic, and this version of cmake doesn't understand "FetchContent".  So I cannot really test it.  That's OK, nobody can really test it until we get a signaling server setup, so even if it's "broken" there isn't anything to break.

I'm going to try switching to use a git submodule instead of having cmake do the git work.  And then cmake will just confirm file existance and tell you to pull in the submodule.  I think that's better for integrating with meson (if we continue to care about meson).  And it means you don't have to have a new version of cmake.

I'll add some tasks for me to do what needs to be done to actually make this stuff usable (dummy signaling service and test case).  Thanks for doing this!

**Pull request merged**